src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs

```diff
@@ -132,102 +145,174 @@ public virtual byte[] GetValue()
+                        // The control is a SortResponseControl. The structure of its value is described as a SortResult in RFC 2891.
+                        ResultCode result;
+                        string attribute = null;
 
-                        result = (int)o[0];
-                    }
+                        AsnDecoder.ReadSequence(asnSpan, AsnEncodingRules.BER, out int sequenceContentOffset, out int sequenceContentLength, out int bytesConsumed);
+                        ThrowUnless(sequenceContentLength > 0);
+                        ThrowUnless(bytesConsumed == asnSpan.Length);
+                        asnSpan = asnSpan.Slice(sequenceContentOffset, sequenceContentLength);
 
-                    SortResponseControl sort = new SortResponseControl((ResultCode)result, attribute, controls[i].IsCritical, controls[i].GetValue());
-                    controls[i] = sort;
-                }
-                else if (controls[i].Type == "2.16.840.1.113730.3.4.10")
-                {
-                    // The control is a VlvResponseControl.
-                    int position;
-                    int count;
-                    int result;
-                    byte[] context = null;
-                    object[] o = BerConverter.TryDecode("{iieO}", value, out bool decodeSucceeded);
-
-                    if (decodeSucceeded)
-                    {
-                        Debug.Assert(o != null && o.Length == 4);
-                        position = (int)o[0];
-                        count = (int)o[1];
-                        result = (int)o[2];
-                        context = (byte[])o[3];
+                        result = AsnDecoder.ReadEnumeratedValue<ResultCode>(asnSpan, AsnEncodingRules.BER, out bytesConsumed);
+                        asnSpan = asnSpan.Slice(bytesConsumed);
+
+                        // If present, the remaining bytes in the control are expected to be an octet string.
+                        if (!asnSpan.IsEmpty)
+                        {
+                            // Attribute name is optional: AD for example never returns attribute name
+                            scoped Span<byte> attributeNameBuffer;
+
+                            if (asnSpan.Length <= AttributeNameStackAllocationThreshold)
+                            {
+                                asnReadSuccessful = AsnDecoder.TryReadOctetString(asnSpan, attributeNameScratchSpace,
+                                    AsnEncodingRules.BER, out bytesConsumed, out int octetStringLength, SortResponseControl.AttributeNameTag);
+                                Debug.Assert(asnReadSuccessful);
+                                attributeNameBuffer = attributeNameScratchSpace.Slice(0, octetStringLength);
+                            }
+                            else
+                            {
+                                attributeNameBuffer = AsnDecoder.ReadOctetString(asnSpan, AsnEncodingRules.BER, out bytesConsumed);
```

Comment:
Why is `SortResponseControl.AttributeNameTag` specified in above for `expectedTag ` in the call to `TryReadOctetString()` but not here in  `ReadOctetString()`?

Also there is no test coverage for the `else` here.

Reply:
Apologies - I'd missed this code path. I've added the test coverage, verified that it fails with the expected error and corrected the bug. There are no other paths with the same problem.

Reply:
Thanks